File: docs/lib/catalogue_service.js
import { Page } from "./page.js";
import { Result } from "./result.js";
import { QueryStringSerializer } from "./querystring_serializer.js";
import { CatalogueConnector } from "./abstractions/catalogue_connector.js";
import { ResultUtils } from "./utils/result_utils.js";

/**
 * @template TItem 
 * @template TFilters
 * @implements {CatalogueConnector}
 */
class CatalogueService {

    /** @type {(json: {}) => TItem} */
    #creator;

    /** @type {(input: RequestInfo | URL, init?: RequestInit | undefined) => Promise<Response>} */
    #fetch;

    /** @type {string} */
    #uri;

    /** @returns {(input: RequestInfo | URL, init?: RequestInit | undefined) => Promise<Response>} */
    get fetch() { return this.#fetch; }

    /** @returns {string} */
    get uri() { return this.#uri; }

    /** @returns {(json: {}) => TItem} */
    get creator() { return this.#creator; }

    /** @returns {boolean} */
    get available() { return navigator.onLine; }

    /**
     * @param {TFilters?} filters 
     * @param {number} pageIndex 
     * @param {number} pageSize 
     * @returns {Promise<Result<Page<TItem>?>>}
     */
    async queryAsync(filters = null, pageIndex = 0, pageSize = DefaultChunkSize) {
        let uri = `${this.#uri}${QueryStringSerializer.serialize(filters)}&pageIndex=${pageIndex}&pageSize=${pageSize}`;
        let response = await this.#fetch(uri, {
            method: "GET"
        });
        let result = await ResultUtils.handleResult(response);
        if (this.#creator && result.data) result.data.items = result.data.items.map(x => this.#creator(x));
        return result;
    }

    /**
     * @param {string} id 
     * @returns {Promise<Result<TItem?>>}
     */
    async getAsync(id) {
        let uri = `${this.#uri}/${id}`;
        let response = await this.#fetch(uri, {
            method: "GET"
        });
        let result = await ResultUtils.handleResult(response);
        if (this.#creator && result.data) result.data = this.#creator(result.data);
        return result;
    }

    /**
     * @param {TItem} item 
     * @returns {Promise<Result<TItem?>>}
     */
    async insertAsync(item) {
        let uri = `${this.#uri}`;
        let response = await this.#fetch(uri, {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify(item)
        });
        let result = await ResultUtils.handleResult(response);
        if (this.#creator && result.data) result.data = this.#creator(result.data);
        return result;
    }

    /**
     * @param {string} id 
     * @param {TItem} item 
     * @returns {Promise<Result<TItem?>>}
     */
    async updateAsync(id, item) {
        let uri = `${this.#uri}/${id}`;
        let response = await this.#fetch(uri, {
            method: "PUT",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify(item)
        });
        let result = await ResultUtils.handleResult(response);
        if (this.#creator && result.data) result.data = this.#creator(result.data);
        return result;
    }

    /**
     * @param {string} id 
     * @returns {Promise<Result<TItem?>>}
     */
    async deleteAsync(id) {
        let uri = `${this.#uri}/${id}`;
        let response = await this.#fetch(uri, {
            method: "DELETE"
        });
        let result = await ResultUtils.handleResult(response);
        if (this.#creator && result.data) result.data = this.#creator(result.data);
        return result;
    }

    /**
     * @param {(json: TItem) => TItem} creator
     * @param {{(input: RequestInfo | URL, init?: RequestInit | undefined) => Promise<Response>}} fetch 
     * @param {string} uri 
     */
    constructor(creator, fetch, uri) {
        this.#creator = creator;
        this.#fetch = fetch;
        this.#uri = uri;
    }

}

export {
    CatalogueService
}

File: Client/CatalogueService.cs
using System.Net.Http.Json;
using System.Net.NetworkInformation;
using KolibSoft.Catalogue.Client.Utils;
using KolibSoft.Catalogue.Core;
using KolibSoft.Catalogue.Core.Abstractions;
using static KolibSoft.Catalogue.Core.Constants;

namespace KolibSoft.Catalogue.Client;

public class CatalogueService<TItem, TFilters> : ICatalogueConnector<TItem, TFilters>
{

    public HttpClient HttpClient { get; }
    public string Uri { get; }
    public virtual bool Available => NetworkInterface.GetAllNetworkInterfaces().Any(x => x.OperationalStatus == OperationalStatus.Up && x.NetworkInterfaceType != NetworkInterfaceType.Loopback);

    public virtual async Task<Result<Page<TItem>?>> QueryAsync(TFilters? filters = default, int pageIndex = 0, int pageSize = DefaultChunkSize)
    {
        var uri = $"{Uri}{QueryStringSerializer.Serialize(filters)}&pageIndex={pageIndex}&pageSize={pageSize}";
        var response = await HttpClient.GetAsync(uri);
        var result = await response.HandleResult<Page<TItem>?>();
        return result;
    }

    public virtual async Task<Result<TItem?>> GetAsync(Guid id)
    {
        var uri = $"{Uri}/{id}";
        var response = await HttpClient.GetAsync(uri);
        var result = await response.HandleResult<TItem?>();
        return result;
    }

    public virtual async Task<Result<TItem?>> InsertAsync(TItem item)
    {
        var uri = $"{Uri}";
        var response = await HttpClient.PostAsJsonAsync(uri, item);
        var result = await response.HandleResult<TItem?>();
        return result;
    }

    public virtual async Task<Result<TItem?>> UpdateAsync(Guid id, TItem item)
    {
        var uri = $"{Uri}/{id}";
        var response = await HttpClient.PutAsJsonAsync(uri, item);
        var result = await response.HandleResult<TItem?>();
        return result;
    }

    public virtual async Task<Result<TItem?>> DeleteAsync(Guid id)
    {
        var uri = $"{Uri}/{id}";
        var response = await HttpClient.DeleteAsync(uri);
        var result = await response.HandleResult<TItem?>();
        return result;
    }

    public CatalogueService(HttpClient httpClient, string uri)
    {
        HttpClient = httpClient;
        Uri = uri;
    }

}